### PR TITLE
Update metadata on replica recovery

### PR DESCRIPTION
**Pull request by tavplubix**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Bug Fix

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed a bug in recovery of staled `ReplicatedMergeTree` replica. Some metadata updates could be ignored by staled replica if `ALTER` query was executed during downtime of the replica.

### CODE REVIEW

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 1]
```diff
@@ -2576,6 +2576,65 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
...
     }
 
+    /// We should do it after copying queue, because some ALTER_METADATA entries can be lost otherwise.
+    Int32 source_metadata_version = parse<Int32>(zookeeper->get(source_path + "/metadata_version"));
+    if (metadata_version != source_metadata_version)
+    {
+        /// Our metadata it not up to date with source replica metadata.
+        /// Metadata is updated by ALTER_METADATA entries, but some entries are probably cleaned up from the log.
+        /// It's also possible that some newer ALTER_METADATA entries are present in source_queue list,
+        /// and source replica are executing such entry right not (or had executed recently).
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:2586]:
```suggestion
        /// and source replica are executing such entry right now (or had executed recently).
```

File: src/Storages/MergeTree/ReplicatedMergeTreeLogEntry.h [hunk 2]
```diff
@@ -121,6 +121,7 @@ struct ReplicatedMergeTreeLogEntryData
     int alter_version = -1; /// May be equal to -1, if it's normal mutation, not metadata update.
 
     /// only ALTER METADATA command
+    /// TODO Seems like it's never used
```
Comment (alesapin) [inline src/Storages/MergeTree/ReplicatedMergeTreeLogEntry.h:124]: Yes, it was a mistake...

File: src/Storages/StorageReplicatedMergeTree.cpp [hunk 3]
```diff
@@ -2576,6 +2576,65 @@ void StorageReplicatedMergeTree::cloneReplica(const String & source_replica, Coo
...
+        dummy_alter.metadata_str = source_metadata;
+        dummy_alter.columns_str = source_columns;
+        dummy_alter.alter_version = source_metadata_version;
+        dummy_alter.create_time = time(nullptr);
+        /// We don't really know if it has mutation or not, but it doesn't matter.
+        dummy_alter.have_mutation = false;
+
+        zookeeper->create(replica_path + "/queue/queue-", dummy_alter.toString(), zkutil::CreateMode::PersistentSequential);
+
+        /// TODO Do we also need to do something with mutation_pointer? Seems like we don't, but it's not clear from the code.
```
Comment (alesapin) [inline src/Storages/StorageReplicatedMergeTree.cpp:2635]: Definitely yes, we shouldn't do anything here, but our mutation pointer must be equal with the source replica's pointer?

**Review by alesapin — APPROVED** [review]

File: src/Storages/StorageReplicatedMergeTree.cpp (hunk 3 above)
Comment (tavplubix) [reply src/Storages/StorageReplicatedMergeTree.cpp:2635]: No, it may be less than mutation pointer of the source replica

**Pull request merged**